### PR TITLE
Add analyser for inconsistent access on bollard/bus_trap vs the highway they're on

### PR DESCRIPTION
**Pull request by Famlam**
Test for inconsistencies between vehicles allowed on a highway and vehicles allowed to pass a barrier on that highway. Only for bollard and bus_trap as they have access/motor_vehicle=no by default
- Also fix the detail text of class 1 (there's no test for `destination`)
- Only test for the vehicle types that are relatively common to have exceptions on a certain type of barrier

I went for a fairly conservative approach (e.g. ignoring all end nodes) to limit the chance of false positives. Also I only check if the vehicle type tag has been set on the bollard, so that tagging like (highway: `moped=designated` + bollard: `moped=yes` is considered valid), and non-passable barriers can still be explicitly tagged as `no` (like a bollard with a traffic sign that forbids passage on that specific point, even though the road allows it)

Results for Flevoland (NL) below as example:
```xml

<error class="2">
	<way id="1106078888" version="1" user="A67-A67">
		<tag k="mofa" v="designated" />
		<tag k="name" v="Cirkelpad" />
		<tag k="moped" v="designated" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="paving_stones" />
		<tag k="traffic_sign" v="NL:G12a" />
		<tag k="zone:traffic" v="NL:rural" />
	</way>
	<node id="2338721604" lat="52.3120854" lon="5.2988076" version="1" user="cartinus">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3120854" lon="5.2988076" />
	<text lang="en" value="Inconsistent mofa access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="44722600" version="13" user="emvee">
		<tag k="mofa" v="yes" />
		<tag k="name" v="Volksdanspad" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="6053233567" lat="52.3632335" lon="5.2511906" version="2" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3632335" lon="5.2511906" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="24430521" version="3" user="emvee">
		<tag k="mofa" v="yes" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="9704081249" lat="52.382670700000006" lon="5.2040398" version="1" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.382670700000006" lon="5.2040398" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="7024997" version="8" user="Geim">
		<tag k="mofa" v="yes" />
		<tag k="name" v="Bodestraat" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="9831508297" lat="52.372179900000006" lon="5.2230675" version="1" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.372179900000006" lon="5.2230675" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="169566380" version="6" user="emvee">
		<tag k="mofa" v="yes" />
		<tag k="name" v="John Gielgudpad" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="9831511817" lat="52.3715005" lon="5.2371165" version="1" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3715005" lon="5.2371165" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="169566380" version="6" user="emvee">
		<tag k="mofa" v="yes" />
		<tag k="name" v="John Gielgudpad" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="252631917" lat="52.3713727" lon="5.237246600000001" version="4" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3713727" lon="5.237246600000001" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="44722600" version="13" user="emvee">
		<tag k="mofa" v="yes" />
		<tag k="name" v="Volksdanspad" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="252542442" lat="52.363102700000006" lon="5.2513109" version="3" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.363102700000006" lon="5.2513109" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="695789013" version="5" user="user_242798">
		<tag k="mofa" v="designated" />
		<tag k="name" v="Gooimeerdijk-West" />
		<tag k="moped" v="designated" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="asphalt" />
		<tag k="traffic_sign" v="NL:G12a" />
	</way>
	<node id="2339461136" lat="52.3293867" lon="5.152538000000001" version="2" user="Marco Nell">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3293867" lon="5.152538000000001" />
	<text lang="en" value="Inconsistent mofa access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="24430521" version="3" user="emvee">
		<tag k="mofa" v="yes" />
		<tag k="moped" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="9704081248" lat="52.3825957" lon="5.2044167" version="1" user="emvee">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3825957" lon="5.2044167" />
	<text lang="en" value="Inconsistent mofa access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="741886077" version="10" user="A67-A67">
		<tag k="mofa" v="designated" />
		<tag k="name" v="Valleipad" />
		<tag k="moped" v="no" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="concrete" />
		<tag k="est_width" v="2.5" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="7577777767" lat="52.3276132" lon="5.333061900000001" version="1" user="PeeWee32">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3276132" lon="5.333061900000001" />
	<text lang="en" value="Inconsistent mofa access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="741886075" version="11" user="A67-A67">
		<tag k="mofa" v="designated" />
		<tag k="name" v="Valleipad" />
		<tag k="moped" v="no" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="concrete" />
		<tag k="est_width" v="2.5" />
		<tag k="traffic_sign" v="NL:G11" />
	</way>
	<node id="7577777768" lat="52.32755" lon="5.332948900000001" version="1" user="PeeWee32">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.32755" lon="5.332948900000001" />
	<text lang="en" value="Inconsistent mofa access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="1106078888" version="1" user="A67-A67">
		<tag k="mofa" v="designated" />
		<tag k="name" v="Cirkelpad" />
		<tag k="moped" v="designated" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="paving_stones" />
		<tag k="traffic_sign" v="NL:G12a" />
		<tag k="zone:traffic" v="NL:rural" />
	</way>
	<node id="2338721604" lat="52.3120854" lon="5.2988076" version="1" user="cartinus">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3120854" lon="5.2988076" />
	<text lang="en" value="Inconsistent moped access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="23245550" version="5" user="PeeWee32">
		<tag k="moped" v="designated" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="paving_stones" />
		<tag k="traffic_sign" v="NL:G12a" />
	</way>
	<node id="247660479" lat="52.43448720000001" lon="5.257454500000001" version="3" user="PeeWee32">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.43448720000001" lon="5.257454500000001" />
	<text lang="en" value="Inconsistent moped access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="695789013" version="5" user="user_242798">
		<tag k="mofa" v="designated" />
		<tag k="name" v="Gooimeerdijk-West" />
		<tag k="moped" v="designated" />
		<tag k="oneway" v="no" />
		<tag k="highway" v="cycleway" />
		<tag k="surface" v="asphalt" />
		<tag k="traffic_sign" v="NL:G12a" />
	</way>
	<node id="2339461136" lat="52.3293867" lon="5.152538000000001" version="2" user="Marco Nell">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.3293867" lon="5.152538000000001" />
	<text lang="en" value="Inconsistent moped access: 'designated' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="356907761" version="11" user="CartoKees">
		<tag k="lit" v="yes" />
		<tag k="name" v="De Promesse" />
		<tag k="oneway" v="no" />
		<tag k="bicycle" v="yes" />
		<tag k="highway" v="pedestrian" />
		<tag k="surface" v="paving_stones" />
		<tag k="maxspeed" v="10" />
		<tag k="emergency" v="yes" />
		<tag k="motor_vehicle:conditional" v="delivery @ (Mo-We 06:00-11:00,18:00-23:00;Th 06:00-11:00;Fr 06:00-11:00,18:00-23:00)" />
	</way>
	<node id="7256215884" lat="52.5089029" lon="5.4746785000000004" version="2" user="Crashtender">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.5089029" lon="5.4746785" />
	<text lang="en" value="Inconsistent emergency access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="777566544" version="6" user="CartoKees">
		<tag k="lit" v="yes" />
		<tag k="name" v="De Promesse" />
		<tag k="oneway" v="no" />
		<tag k="bicycle" v="yes" />
		<tag k="highway" v="pedestrian" />
		<tag k="surface" v="paving_stones" />
		<tag k="maxspeed" v="10" />
		<tag k="emergency" v="yes" />
		<tag k="motor_vehicle:conditional" v="delivery @ (Mo-We 06:00-11:00,18:00-23:00;Th 06:00-11:00;Fr 06:00-11:00,18:00-23:00)" />
	</way>
	<node id="7256215876" lat="52.5079695" lon="5.474628500000001" version="1" user="Crashtender">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.5079695" lon="5.474628500000001" />
	<text lang="en" value="Inconsistent emergency access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="777548164" version="6" user="Reisinformatiegroep">
		<tag k="lit" v="yes" />
		<tag k="name" v="De Promesse" />
		<tag k="oneway" v="no" />
		<tag k="bicycle" v="yes" />
		<tag k="highway" v="pedestrian" />
		<tag k="surface" v="paving_stones" />
		<tag k="maxspeed" v="10" />
		<tag k="emergency" v="yes" />
		<tag k="motor_vehicle:conditional" v="delivery @ (Mo-We 06:00-11:00,18:00-23:00;Th 06:00-11:00;Fr 06:00-11:00,18:00-23:00)" />
	</way>
	<node id="7256215880" lat="52.5086153" lon="5.4744212" version="2" user="Crashtender">
		<tag k="barrier" v="bollard" />
	</node>
	<location lat="52.5086153" lon="5.4744212" />
	<text lang="en" value="Inconsistent emergency access: 'yes' on highway, not set on barrier" />
</error>
<error class="2">
	<way id="43084763" version="5" user="A67-A67">
		<tag k="bus" v="yes" />
		<tag k="foot" v="yes" />
		<tag k="name" v="Kempenaar 14" />
		<tag k="moped" v="yes" />
		<tag k="bicycle" v="yes" />
		<tag k="highway" v="residential" />
		<tag k="maxspeed" v="30" />
		<tag k="motor_vehicle" v="no" />
	</way>
	<node id="465698105" lat="52.5147989" lon="5.4660395" version="5" user="Martijn-VR-Gooi en Vechtstreek">
		<tag k="foot" v="yes" />
		<tag k="barrier" v="bus_trap" />
	</node>
	<location lat="52.5147989" lon="5.4660395" />
	<text lang="en" value="Inconsistent moped access: 'yes' on highway, not set on barrier" />
</error>
```

Some other results attached: [zip](https://github.com/osm-fr/osmose-backend/files/9946245/results.zip)

p.s. slightly offtopic, but the name of the class / python file is a bit misleading: none of the classes does anything with area's

### CODE REVIEW

**Comment by Famlam** [comment]
The lint errors seem unrelated to this PR

**Comment by Famlam** [comment]
@frodrigo I assume you haven't had time yet to review this?
Just 'pinging' in case you missed it :)

File: analysers/analyser_osmosis_highway_area_access.py [hunk 1]
```diff
@@ -51,24 +51,120 @@
...
+    NOT ways.is_area
+  LEFT JOIN highway_ends ON
+    ways.id = highway_ends.id AND
+    barrier.id = highway_ends.nid
+WHERE
+  barrier.tags != ''::hstore AND
+  barrier.tags?'barrier' AND
+  barrier.tags->'barrier' = '{barriertype}' AND
+  /* Default of bollard is access=no / bicycle=foot=yes; default of bus_trap is motor_vehicle=no / psv=foot=bicycle=yes.
+  Hence, the below three lines should cover all cases as long as we don't test for any non-motor_vehicle or vehicles under psv */
```
Comment (frodrigo) [inline analysers/analyser_osmosis_highway_area_access.py:77]:
```suggestion
  -- Default of bollard is access=no / bicycle=foot=yes; default of bus_trap is motor_vehicle=no / psv=foot=bicycle=yes.
  -- Hence, the below three lines should cover all cases as long as we don't test for any non-motor_vehicle or vehicles under psv
```

File: analysers/analyser_osmosis_highway_area_access.py [hunk 2]
```diff
@@ -51,24 +51,120 @@
...
+  barrier.id AS id,
+  barrier.tags AS tags,
+  ways.id as wid,
+  ST_AsText(barrier.geom) AS geom
+FROM
+  nodes AS barrier
+  JOIN way_nodes ON
+    way_nodes.node_id = barrier.id
+  JOIN highways AS ways ON
+    ways.id = way_nodes.way_id AND
```
Comment (frodrigo) [inline analysers/analyser_osmosis_highway_area_access.py:66]: It is not intuitive, but this is faster.

```suggestion
  JOIN highways AS ways ON
    barrier.id = ANY (ways.nodes) AND
```

File: analysers/analyser_osmosis_highway_area_access.py [hunk 3]
```diff
@@ -51,24 +51,120 @@
...
+    NOT ways.is_construction AND
+    NOT ways.is_area
+  LEFT JOIN highway_ends ON
+    ways.id = highway_ends.id AND
+    barrier.id = highway_ends.nid
```
Comment (frodrigo) [inline analysers/analyser_osmosis_highway_area_access.py:71]: There is faster way to test if the node is at the ends that not require a left join. Just look into the nodes array.

File: analysers/analyser_osmosis_highway_area_access.py [hunk 4]
```diff
@@ -51,24 +51,120 @@
   )
 """
 
+sql20 = """
+CREATE TEMP TABLE {barriertype} AS
+SELECT
+  barrier.id AS id,
+  barrier.tags AS tags,
+  ways.id as wid,
+  ST_AsText(barrier.geom) AS geom
```
Comment (frodrigo) [inline analysers/analyser_osmosis_highway_area_access.py:60]: Make `ST_AsText` only on sql21, on results.

File: analysers/analyser_osmosis_highway_area_access.py [hunk 5]
```diff
@@ -51,24 +51,120 @@
...
+    barrier.wid,
+    barrier.id,
+    barrier.geom,
+    highway.tags->'{vehicle}'
+FROM
+    {barriertype} AS barrier
+    JOIN highways AS highway ON
+      highway.id = barrier.wid AND
+      highway.tags?'{vehicle}' AND
+      NOT highway.tags->'{vehicle}' IN ('no', 'use_sidepath', 'unknown')
```
Comment (frodrigo) [inline analysers/analyser_osmosis_highway_area_access.py:95]: Ouput `tags?'{vehicle}'` in sql20 and avoid a join here.

**Comment by frodrigo** [comment]
> @frodrigo I assume you haven't had time yet to review this? Just 'pinging' in case you missed it :)

Thank for remind. Just missing time these days.

**Comment by Famlam** [comment]
Thank you for the review! I've implemented the changes, but there seems to be some performance issues. Below the time stamps for the runs before (`_old` suffix) and after (no suffix) the commit. This is very reproducible also on other countries. I'll dive into it later which of the changes causes this slowdown...

```
2022-11-28 11:57:05 netherlands_flevoland : osmosis_highway_area_access_old
2022-11-28 11:57:05   run osmosis all analyser Analyser_Osmosis_HighwayAreaAccess
2022-11-28 11:57:05   requires table highway_ends
2022-11-28 11:57:05   requires table highways
2022-11-28 11:57:09   analyser_osmosis_highway_area_access_old.py:127 sql <<<<<<<<<< 23 seconds before the last commit
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:128 sql
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:130 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:130 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:130 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:130 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:130 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:134 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:134 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:134 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:134 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:134 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:134 xml generation
2022-11-28 11:57:32   analyser_osmosis_highway_area_access_old.py:139 xml generation
2022-11-28 11:57:34 netherlands_flevoland : osmosis_highway_area_access
2022-11-28 11:57:34   run osmosis all analyser Analyser_Osmosis_HighwayAreaAccess
2022-11-28 11:57:34   analyser_osmosis_highway_area_access.py:120 sql <<<<<<<<<< 82 seconds since the last commit
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:121 sql
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:123 xml generation
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:123 xml generation
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:123 xml generation
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:123 xml generation
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:123 xml generation
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:127 xml generation
2022-11-28 11:58:56   analyser_osmosis_highway_area_access.py:127 xml generation
2022-11-28 11:58:57   analyser_osmosis_highway_area_access.py:127 xml generation
2022-11-28 11:58:57   analyser_osmosis_highway_area_access.py:127 xml generation
2022-11-28 11:58:57   analyser_osmosis_highway_area_access.py:127 xml generation
2022-11-28 11:58:57   analyser_osmosis_highway_area_access.py:127 xml generation
2022-11-28 11:58:57   analyser_osmosis_highway_area_access.py:132 xml generation
```

**Comment by Famlam** [comment]
The slowdown seems to be related to https://github.com/osm-fr/osmose-backend/pull/1642#discussion_r1032942255
When I revert that part only, I get: 
```
2022-11-28 13:43:01 netherlands_flevoland : osmosis_highway_area_access
2022-11-28 13:43:01   run osmosis all analyser Analyser_Osmosis_HighwayAreaAccess
2022-11-28 13:43:01   requires table highways
2022-11-28 13:43:03   analyser_osmosis_highway_area_access.py:122 sql
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:123 sql
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:125 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:125 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:125 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:125 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:125 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:129 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:129 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:129 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:129 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:129 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:129 xml generation
2022-11-28 13:43:04   analyser_osmosis_highway_area_access.py:134 xml generation
```

I'll verify if the results are the same later this week.

File: analysers/analyser_osmosis_highway_area_access.py (hunk 5 above)
Comment (Famlam) [reply analysers/analyser_osmosis_highway_area_access.py]: Thanks! I did implement it slightly different by exporting all way tags, so I can avoid having to put sql20 in the loop (or having to check for all tags one by one here), but indeed much better without the join

File: analysers/analyser_osmosis_highway_area_access.py (hunk 2 above)
Comment (Famlam) [reply analysers/analyser_osmosis_highway_area_access.py:67]: Unfortunately this seems to cause a significant slowdown: about a factor 4 for medium-sized countries like netherlands_flevoland, to a factor 300 for a huge country like the (non-split-up) belgium_flanders. Maybe it's only faster if there's only few ways/nodes to compare against? In this sql query it's practically called for every highway way and node as there's no filtering preceding it.

**Comment by frodrigo** [comment]
> The slowdown seems to be related to https://github.com/osm-fr/osmose-backend/pull/1642#discussion_r1032942255

Strange. To be checked. But I think all this kind of joins are done like this.
Maybe it depends of the size of the extracts.

**Comment by Famlam** [comment]
Yeah, probably, because for bigger extracts the time difference was much more severe than for smaller countries. 

Anything you'd like me to do on this PR?

**Comment by frodrigo** [comment]
> Yeah, probably, because for bigger extracts the time difference was much more severe than for smaller countries. It also seems to be very specific for this specific SQL query: if I just run
> 
> ```sql
> CREATE TEMP TABLE x AS
> SELECT
>   barrier.id
> FROM
>   nodes AS barrier
>   JOIN way_nodes ON
>     way_nodes.node_id = barrier.id
>   JOIN highways AS ways ON
>     ways.id = way_nodes.way_id
> WHERE
>   barrier.id != ways.nodes[1] AND barrier.id != ways.nodes[array_length(ways.nodes,1)]
> ```
> 
> vs
> 
> ```sql
> CREATE TEMP TABLE x AS
> SELECT
>   barrier.id
> FROM
>   nodes AS barrier
>   JOIN highways AS ways ON
>     barrier.id = ANY(ways.nodes)
> WHERE
>   barrier.id != ways.nodes[1] AND barrier.id != ways.nodes[array_length(ways.nodes,1)]
> ```
> 
> Then the difference is very small, regardless of country size.
> 
> Anything you'd like me to do on this PR?

On this the 1-join is faster than the 2-join.

ON france_aquitaine_gironde
Explain
```
2-join: Hash Join  (cost=407790.63..1248804.68 rows=13377415 width=8)
1-join: Gather     (cost=  1000.43.. 749046.03 rows= 1747869 width=8)
```

2-join (first run still slower due to fetch from disk)
```
Time: 24569.655 ms (00:24.570)
Time: 20417.106 ms (00:20.417)
Time: 19790.026 ms (00:19.790)
```
1-join
```
Time: 5586.955 ms (00:05.587)
Time: 4738.014 ms (00:04.738)
Time: 4611.905 ms (00:04.612)
```

On netherlands_flevoland
Explain
```
2-join: Hash Join  (cost=128771.81..379350.50 rows=3978408 width=8)
1-join: Gather     (cost=  1000.43..297685.37 rows= 726358 width=8)
```

2-join
```
Time: 5472.510 ms (00:05.473)
Time: 4965.617 ms (00:04.966)
Time: 5262.646 ms (00:05.263)
```

1-join
```
Time: 1284.439 ms (00:01.284)
Time: 1162.244 ms (00:01.162)
Time: 1147.781 ms (00:01.148)
```

**Comment by frodrigo** [comment]
On original query, 1-join raise an issue with query planner. It try to uses `idx_nodes_tags`, but it is a bad plan.

We can disable this plan by commenting `barrier.tags?'barrier'`.

Once it is done. 1-join and 2-join query are close the same, less than 1s for netherlands_flevoland, and about 3 s for france_aquitaine_gironde.

2-join is even faster in this case. I no deep into to get why.

Use 1-join or 2-join, as you wish. But comment the close introducing bad plan.

```sql
CREATE TEMP TABLE bollard AS
--explain
SELECT
  barrier.id AS id,
  barrier.tags AS tags,
  ways.id AS wid,
  barrier.geom AS geom,
  ways.tags AS waytags
FROM
  nodes AS barrier

 JOIN way_nodes ON
    way_nodes.node_id = barrier.id
  JOIN highways AS ways ON
    barrier.id = ANY(ways.nodes) AND
    ways.id = way_nodes.way_id AND

--   JOIN highways AS ways ON
--     barrier.id = ANY(ways.nodes) AND

    NOT ways.is_construction AND
    NOT ways.is_area
WHERE
  barrier.tags != ''::hstore AND
  -- barrier.tags?'barrier' AND -- commented to avoid planner to try to use idx_nodes_tags, build bad plan with it.
  barrier.tags->'barrier' = 'bollard' AND
  -- Default of bollard is access=no / bicycle=foot=yes; default of bus_trap is motor_vehicle=no / psv=foot=bicycle=yes.
  -- Hence, the below three lines should cover all cases as long as we don't test for any non-motor_vehicle or vehicles under psv
  (NOT barrier.tags?'access' OR barrier.tags->'access' = 'no') AND
  (NOT barrier.tags?'vehicle' OR barrier.tags->'vehicle' = 'no') AND
  (NOT barrier.tags?'motor_vehicle' OR barrier.tags->'motor_vehicle' = 'no') AND
  barrier.id != ways.nodes[1] AND barrier.id != ways.nodes[array_length(ways.nodes,1)] -- Barrier is not an end node
;
```

**Comment by Famlam** [comment]
Thank you very much for the detailed explanation!
I've committed the patch. However, I must say I don't fully understand how to read the `EXPLAIN` queries. I can reproduce your results of being much faster without the  `barrier.tags?'barrier'` line:
<details>
<summary>Click here for EXPLAIN output</summary>

With the 1-join original query, on netherlands_flevoland I get:

```
['Nested Loop  (cost=10.16..8161.23 rows=1 width=156) (actual time=266.526..101533.477 rows=446 loops=1)']
['  Join Filter: ((barrier.id <> ways.nodes[1]) AND (barrier.id <> ways.nodes[array_length(ways.nodes, 1)]) AND (barrier.id = ANY (ways.nodes)))']
['  Rows Removed by Join Filter: 92754730']
['  ->  Bitmap Heap Scan on nodes barrier  (cost=10.16..962.67 rows=1 width=57) (actual time=66.512..974.183 rows=1307 loops=1)']
["        Recheck Cond: ((tags ? 'barrier'::text) AND (tags <> ''::hstore))"]
['        Rows Removed by Index Recheck: 4969']
["        Filter: (((tags -> 'barrier'::text) = 'bollard'::text) AND ((NOT (tags ? 'access'::text)) OR ((tags -> 'access'::text) = 'no'::text)) AND ((NOT (tags ? 'vehicle'::text)) OR ((tags -> 'vehicle'::text) = 'no'::text)) AND ((NOT (tags ? 'motor_vehicle'::text)) OR ((tags -> 'motor_vehicle'::text) = 'no'::text)))"]
['        Rows Removed by Filter: 1802']
['        Heap Blocks: exact=3700']
['        ->  Bitmap Index Scan on idx_nodes_tags  (cost=0.00..10.16 rows=250 width=0) (actual time=64.632..64.633 rows=8078 loops=1)']
["              Index Cond: (tags ? 'barrier'::text)"]
['  ->  Seq Scan on highways ways  (cost=0.00..5069.04 rows=70984 width=165) (actual time=0.011..40.126 rows=70968 loops=1307)']
['        Filter: ((NOT is_construction) AND (NOT is_area))']
['        Rows Removed by Filter: 736']
['Planning Time: 5.025 ms']
['Execution Time: 101534.006 ms']
```

After commenting out `barrier.tags?'barrier'` I get:
```
['Gather  (cost=1000.43..227118.50 rows=293 width=156) (actual time=49.389..1202.179 rows=446 loops=1)']
['  Workers Planned: 2']
['  Workers Launched: 2']
['  ->  Nested Loop  (cost=0.43..226089.20 rows=122 width=156) (actual time=38.517..1177.818 rows=149 loops=3)']
['        ->  Parallel Seq Scan on highways ways  (cost=0.00..4650.77 rows=29577 width=165) (actual time=0.105..90.093 rows=23656 loops=3)']
['              Filter: ((NOT is_construction) AND (NOT is_area))']
['              Rows Removed by Filter: 245']
['        ->  Index Scan using pk_nodes on nodes barrier  (cost=0.43..7.48 rows=1 width=57) (actual time=0.045..0.045 rows=0 loops=70968)']
['              Index Cond: (id = ANY (ways.nodes))']
["              Filter: ((tags <> ''::hstore) AND (id <> (ways.nodes)[1]) AND ((tags -> 'barrier'::text) = 'bollard'::text) AND (id <> (ways.nodes)[array_length(ways.nodes, 1)]) AND ((NOT (tags ? 'access'::text)) OR ((tags -> 'access'::text) = 'no'::text)) AND ((NOT (tags ? 'vehicle'::text)) OR ((tags -> 'vehicle'::text) = 'no'::text)) AND ((NOT (tags ? 'motor_vehicle'::text)) OR ((tags -> 'motor_vehicle'::text) = 'no'::text)))"]
['              Rows Removed by Filter: 6']
['Planning Time: 0.919 ms']
['Execution Time: 1202.318 ms']
```
</details>


But in the original `EXPLAIN` I only see:
```
['        ->  Bitmap Index Scan on idx_nodes_tags  (cost=0.00..10.16 rows=250 width=0) (actual time=64.632..64.633 rows=8078 loops=1)']
["              Index Cond: (tags ? 'barrier'::text)"]
```
64 milliseconds. 
`['Nested Loop  (cost=10.16..8161.23 rows=1 width=156) (actual time=266.526..101533.477 rows=446 loops=1)']`

Do you have any advice on how to interpret the `EXPLAIN` output for me?

**Comment by Famlam** [comment]
Results look good :)

**Pull request merged**